runtime/roots.c

```diff
@@ -0,0 +1,77 @@
+/**************************************************************************/
+/*                                                                        */
+/*                                 OCaml                                  */
+/*                                                                        */
+/*          Xavier Leroy and Damien Doligez, INRIA Rocquencourt           */
+/*                                                                        */
+/*   Copyright 1996 Institut National de Recherche en Informatique et     */
+/*     en Automatique.                                                    */
+/*                                                                        */
+/*   All rights reserved.  This file is distributed under the terms of    */
+/*   the GNU Lesser General Public License version 2.1, with the          */
+/*   special exception on linking described in the file LICENSE.          */
+/*                                                                        */
+/**************************************************************************/
+
+#define CAML_INTERNALS
+
+/* To walk the memory roots for garbage collection */
+
+#include "caml/finalise.h"
+#include "caml/globroots.h"
+#include "caml/major_gc.h"
+#include "caml/memory.h"
+#include "caml/minor_gc.h"
+#include "caml/misc.h"
+#include "caml/mlvalues.h"
+#include "caml/roots.h"
+#include "caml/major_gc.h"
+#include "caml/shared_heap.h"
+#include "caml/fiber.h"
+
+#ifdef NATIVE_CODE
+#include "caml/stack.h"
+/* Communication with [caml_start_program] and [caml_call_gc]. */
+
+/* The global roots.
+   FIXME: These should be promoted, and not scanned here.
+   FIXME: caml_globals_inited makes assumptions about store ordering.
+   XXX KC : What to do here?
+*/
+
+intnat caml_globals_inited = 0;
```

Comment:
We should discuss this point, indeed.  However I'm a bit lost in the code right now: where is the scanning of native-code globals now that you've deleted `runtime/roots_nat.c` ?  At any rate, the idea is to scan only those globals that have not been scanned before and are already initialized.  Otherwise, since there can be many globals, scanning takes significantly more time.  

Concerning store ordering: I was under the impression that roots are scanned only when all mutators are stopped and crossed appropriate memory barriers so that all of their writes are visible to the GC.  In which case we should be all right, no?

Reply:
These comments are out of date, and I shall remove them. 

> I was under the impression that roots are scanned only when all mutators are stopped and crossed appropriate memory barriers so that all of their writes are visible to the GC. In which case we should be all right, no?

This is correct. In multicore today, only one domain scans the global roots and there is a barrier at the end of the global root marking before the mutators can run. See https://github.com/ocaml-multicore/ocaml/blob/multicore-pr/runtime/major_gc.c#L1155-L1158. 

`scan_native_globals` is now in `globroots.c`: https://github.com/ocaml-multicore/ocaml/blob/multicore-pr/runtime/globroots.c#L186. The global roots management was discussed in https://github.com/ocaml-multicore/ocaml-multicore/issues/424, and many of the issues have been fixed, and further improvements have been listed here: https://github.com/ocaml-multicore/ocaml-multicore/issues/578. 

As you have rightly pointed out, multicore doesn't yet optimise for globals already scanned (multicore doesn't have the `caml_globals_scanned` variable). This should be easy to fix as exactly one domain scans the global roots.

Reply:
> ~These comments are out of date, and I shall remove them.~

If I am reading the code correctly, then multicore does not oldify the globals in `caml_globals` table, which trunk does. See https://github.com/ocaml-multicore/ocaml/blob/multicore-pr/runtime/globroots.c#L242 in multicore and compare that with https://github.com/ocaml/ocaml/blob/trunk/runtime/roots_nat.c#L258-L268 on trunk. Why is this safe? Perhaps @sadiqj @ctk21 will know.

Reply:
Multicore OCaml uses `caml_initialize` to store the value of a definition at the top level of a compilation unit into the corresponding field of the global data for the compilation unit.  Hence, if the value points to the young heap, it's added to the remembered set, and the next minor GC will treat it from there.

So, this works.  But a downside is to make the initialization code for compilation units even bigger than it is today.   There's a design choice here that we should have discussed during review.

Reply:
As far as I can see, the `caml_globals_inited` optimization about scanning globals is sound in multicore, because all the initialization code for compilation units is executed sequentially on the initial domain.  The initialization code can create new domain as it goes, but these other domains will not modify `caml_globals_inited` nor run initialization code.

Reply:
Thanks for the clarification.

We may revisit this design choice after the merge. The change is fairly easy to implement and benchmark.